### PR TITLE
Add IntervalBucket and tests.

### PR DESCRIPTION
**Pull request by songy23**
`IntervalBucket` is used to implement interval stats. Currently each interval MutableViewData will have by default 5 (N+1) `IntervalBucket`s, and each `IntervalBucket` has a duration which is 1/4 of the total duration. Within each `IntervalBucket`, stats will be aggregated cumulatively. When creating an interval ViewData, stats in all `IntervalBucket`s in the MutableViewData will be combined and aggregated.

### CODE REVIEW

File: core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+final class IntervalBucket {
+
+  private final Timestamp start;
+  private final Duration duration;
+  private final List<Aggregation> aggregations;
+  private final Map<List<TagValue>, List<MutableAggregation>> tagValueAggregationMap =
+      Maps.newHashMap();
+
+  IntervalBucket(Timestamp start, Duration duration, List<Aggregation> aggregations) {
+    checkNotNull(start, "Timestamp");
```
Comment (dinooliva) [inline core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java]: Should this be "Start" rather than "Timestamp"?

Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java]: Right, updated

File: core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  public void testGetStart() {
+    assertThat(new IntervalBucket(START, MINUTE, AGGREGATIONS_V1).getStart()).isEqualTo(START);
+  }
+
+  @Test
+  public void testRecord() {
+    IntervalBucket bucket = new IntervalBucket(START, MINUTE, AGGREGATIONS_V1);
+    List<TagValue> tagValues = Arrays.<TagValue>asList(TagValueString.create("VALUE"));
+    bucket.record(tagValues, 5.0);
+    bucket.record(tagValues, 15.0);
```
Comment (sebright) [inline core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java]: Suggestion: This test would be more thorough if it used multiple tag value combinations.

File: core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  @Test
+  public void preventNullAggregationList() {
+    thrown.expect(NullPointerException.class);
+    new IntervalBucket(START, MINUTE, null);
+  }
+  
+  @Test
+  public void testGetTagValueAggregationMap() {
```
[the remark this answers is not in the document]
Comment (songy23) [reply core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java]: Done.

File: core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java (hunk 2 above)
Comment (songy23) [reply core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java]: Sure, updated.

File: core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+  @Test
+  public void testRecord() {
+    IntervalBucket bucket = new IntervalBucket(START, MINUTE, AGGREGATIONS_V1);
+    List<Object> tagValues = Arrays.<Object>asList(TagValueString.create("VALUE"));
+    bucket.record(tagValues, 5.0);
+    bucket.record(tagValues, 15.0);
+    assertThat(bucket.getTagValueAggregationMap()).containsKey(tagValues);
+    for (MutableAggregation aggregation : bucket.getTagValueAggregationMap().get(tagValues)) {
+      aggregation.match(
```
[the remark this answers is not in the document]
Comment (songy23) [reply core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java]: Good suggestion, updated.

File: core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+   * Returns how much fraction of duration has passed in this IntervalBucket. For example, if this
+   * bucket starts at 10s and has a duration of 20s, and now is 15s, then getFraction() should
+   * return (15 - 10) / 20 = 0.25.
+   *
+   * This IntervalBucket must be current, i.e. the current timestamp must be within
+   * [this.start, this.start + this.duration).
+   */
+  double getFraction(Timestamp now) {
+    Duration elapsedTime = now.subtractTimestamp(start);
+    checkArgument(toMillis(elapsedTime) >= 0 && toMillis(elapsedTime) < toMillis(duration),
```
[the remark this answers is not in the document]
Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/IntervalBucket.java]: I added a TODO for Duration.

File: core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  @Test
+  public void testGetFraction() {
+    Timestamp thirtySecondsAfterStart = Timestamp.create(90, 0);
+    assertThat(
+        new IntervalBucket(START, MINUTE, AGGREGATIONS_V1).getFraction(thirtySecondsAfterStart))
+        .isWithin(TOLERANCE).of(0.5);
+  }
+
+  @Test
+  public void preventCallingGetFractionOnPastBuckets() {
```
[the remark this answers is not in the document]
Comment (songy23) [reply core_impl/src/test/java/io/opencensus/implcore/stats/IntervalBucketTest.java]: Sure, added.

**Comment by bogdandrutu** [comment]
Why did you need to change the gradle config? Please rollback the change if not necessary.

**Comment by songy23** [comment]
Looks like Intellij re-generates gradle config files and by accident I included them in the commit. Thanks for fixing.

**Pull request merged**